### PR TITLE
Embeddings based topic reduction + get rid of ST

### PR DESCRIPTION
**Pull request by dopc**
Moving to here from #11

### CODE REVIEW

File: bertopic/_bertopic.py [hunk 1]
```diff
@@ -860,42 +817,54 @@ def _extract_topics(self, documents: pd.DataFrame):
...
                                              axis=0)
-                topic_embeddings.append(topic_embedding)
+                topics_embeddings.append(topic_embedding)
+        elif method == 'doc':
+            if embeddings is None or predictions is None:
+                raise ValueError("Need embeddings and predictions for \"doc\" topic vectors.")
+
+            for topic_id in topic_list:  # todo: we may iterate with, set(predictions) ?
+                # get each topic's docs' embeddings
+                topic_document_vectors = embeddings[np.where(np.array(predictions) == topic_id)[0]]
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: Where does this 0 come from?

File: bertopic/_bertopic.py [hunk 2]
```diff
@@ -1049,7 +1018,68 @@ def _select_embedding_model(self) -> SentenceTransformer:
...
+        topics, probs = model.fit_transform(docs)
+
+        # Further reduce topics
+        new_topics, new_probs = model.reduce_topics(docs, topics, probs, nr_topics=30)
+        ```
+        """
+        check_is_fitted(self)
+        # be sure that we will return at least 2 topics except outliers(-1)
+        nr_topics = max(nr_topics, 2)
+        self.nr_topics = nr_topics
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]:
```suggestion
        self.nr_topics = max(nr_topics, 2)
```

File: bertopic/_bertopic.py [hunk 3]
```diff
@@ -1058,15 +1087,120 @@ def _reduce_topics(self, documents: pd.DataFrame) -> pd.DataFrame:
...
+        initial_nr_topics = len(self.get_topics())
+
+        topic_vectors = self.topic_embeddings_doc
+        predictions = GaussianMixture(n_components=self.nr_topics,
+                                      random_state=0).fit_predict(topic_vectors)
+
+        merging_mapping = self._get_merging_mapping(predictions)  # mapping old and new topics
+        cluster_mapping = dict()
+
+        if self.mapped_topics is not None:
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: This if statement is always true because not defined mapped_topics's value is set to dict() in the first line.

File: bertopic/_bertopic.py (hunk 1 above)
Comment (zafercavdar) [reply bertopic/_bertopic.py]: It's decided that it will be replaced with `embeddings[np.array(predictions) == topic_id)]`

File: bertopic/_bertopic.py [hunk 4]
```diff
@@ -1058,15 +1087,120 @@ def _reduce_topics(self, documents: pd.DataFrame) -> pd.DataFrame:
...
+            for prev_to_merge, prev_to_merge_into in self.mapped_topics.items():
+                # since we cannot access old topic embeddings, we need to map documents w.r.t. prev. mapping  # noqa
+                documents.loc[documents.Topic == prev_to_merge, "Topic"] = prev_to_merge_into
+
+        for cluster, topics_in_cluster in merging_mapping.items():
+            candidates_for_into = deepcopy(topics_in_cluster)
+
+            topic_to_merge_into = max(candidates_for_into, key=lambda x: self.get_topic_freq(x))
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: We need to make sure that get_topic_freq uses the latest version of df.

File: bertopic/_bertopic.py [hunk 5]
```diff
@@ -1058,15 +1087,120 @@ def _reduce_topics(self, documents: pd.DataFrame) -> pd.DataFrame:
...
+        return documents
+
+    def _reduce_with_hdbscan(self, documents: pd.DataFrame, embeddings: np.ndarray) -> pd.DataFrame:
+        if not self.mapped_topics:
+            self.mapped_topics = dict()
+
+        initial_nr_topics = len(self.get_topics())
+
+        old_predictions = deepcopy(documents.Topic.values)
+        flat_clusterer = HDBSCAN_flat(embeddings, n_clusters=self.nr_topics,
```
Comment (zafercavdar) [inline bertopic/_bertopic.py:1087]: Update hdbscan version in Dockerfile. Check if the old models can be loaded in the new version.

File: bertopic/_bertopic.py [hunk 6]
```diff
@@ -1058,15 +1087,120 @@ def _reduce_topics(self, documents: pd.DataFrame) -> pd.DataFrame:
...
+        processed_diff_count = 0
+        diff_id = 0
+        while processed_diff_count < changed_topic_count:
+            old_topic_label = non_outlier_diff_old[diff_id]
+            new_topic_label = non_outlier_diff_new[diff_id]
+
+            if old_topic_label not in diff_mapping[new_topic_label]:
+                diff_mapping[new_topic_label].append(old_topic_label)
+                processed_diff_count += 1
+            diff_id += 1
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]:
```suggestion
        new_predictions = deepcopy(flat_clusterer.labels_)
        diff_mapping = defaultdict(list)
        
        for old_pred, new_pred in zip(old_predictions, new_predictions):
             if old_pred != new_pred and old_pred != -1 and old_pred not in diff_mapping[new_pred]:
                  diff_mapping[new_pred].append(old_pred)
```

File: bertopic/_bertopic.py [hunk 7]
```diff
@@ -1058,15 +1087,120 @@ def _reduce_topics(self, documents: pd.DataFrame) -> pd.DataFrame:
...
+
+        for new_label, old_labels in diff_mapping.items():
+            candidates_for_into = deepcopy(old_labels)
+            if self.mapped_topics is not None:
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: Unnecessary check.

File: bertopic/_bertopic.py [hunk 8]
```diff
@@ -1058,15 +1087,120 @@ def _reduce_topics(self, documents: pd.DataFrame) -> pd.DataFrame:
...
+                old_tos = self.mapped_topics.keys()
+                for candidate in candidates_for_into:
+                    if candidate in old_tos:
+                        candidates_for_into.remove(candidate)
+
+            topic_to_merge_into = max(candidates_for_into, key=lambda x: old_labels.count(x))
+            old_labels.remove(topic_to_merge_into)  # topics to merge
+
+            for topic_to_merge in old_labels:
+                if topic_to_merge != topic_to_merge_into:
```
Comment (zafercavdar) [inline bertopic/_bertopic.py:1108]: Unnecessary check.

File: bertopic/_bertopic.py [hunk 9]
```diff
@@ -860,42 +817,54 @@ def _extract_topics(self, documents: pd.DataFrame):
...
             # The embeddings var is a single numpy matrix and therefore slicing is necessary to
             # access the words per topic
-            topic_embeddings = []
             for i, topic in enumerate(topic_list):
                 word_importance = [val[1] for val in self.get_topic(topic)]
                 if sum(word_importance) == 0:
                     word_importance = [1 for _ in range(len(self.get_topic(topic)))]
                 topic_embedding = np.average(embeddings[i * n: n + (i * n)],
                                              weights=word_importance,
                                              axis=0)
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: Add a note here about potential issues that may occur in word index-based topic vector extraction.

**Review by zafercavdar — CHANGES_REQUESTED** [review]

File: bertopic/_bertopic.py [hunk 10]
```diff
@@ -393,6 +405,8 @@ def transform(self,
             predictions = self._map_predictions(predictions)
             probabilities = self._map_probabilities(probabilities)
 
+        if probabilities is not None:
+            probabilities = probabilities.round(3)
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: Let's remove this.

File: bertopic/_bertopic.py [hunk 11]
```diff
@@ -312,13 +320,17 @@ def fit_transform(self,
...
+        #     probabilities = self._map_probabilities(probabilities)
 
         predictions = documents.Topic.to_list()
-
-        return predictions, probabilities
+        if self.clustering_method == 'gmm':
+            self.topic_embeddings_doc = self._create_topic_vectors(embeddings=umap_embeddings,
+                                                                   predictions=predictions)
+        if probabilities is not None:
+            probabilities = probabilities.round(3)
```
Comment (zafercavdar) [inline bertopic/_bertopic.py]: Let's remove this.

**Review by zafercavdar — CHANGES_REQUESTED** [review]

**Review by zafercavdar — APPROVED** [review]
